Web public streams

clarammdantas opened the pull request:
<!-- What's this PR for?  (Just a link to an issue is fine.) -->
Make it possible to create web public streams. Guest users can browse and subscribe to them. This uses the prototype on #14491 

Fixes: #14490 

**Testing Plan:** <!-- How have you tested? -->
Manual and automated tests.


**GIFs or Screenshots:** <!-- If a UI change.  See:
  https://zulip.readthedocs.io/en/latest/tutorials/screenshot-and-gif-software.html
  -->

Create `Web Public` stream.

![Captura de tela de 2020-04-18 18-11-06](https://user-images.githubusercontent.com/15564640/79671405-68218880-81a0-11ea-9c9e-26514ea44133.png)


Searching/Subscribing to a `Web Public` stream.
![Captura de tela de 2020-04-18 18-11-06](https://media.giphy.com/media/eKmwj7sspwV0HEDMSR/giphy.gif)


<!-- Also be sure to make clear, coherent commits:
  https://zulip.readthedocs.io/en/latest/contributing/version-control.html
  -->

timabbott commented:
@clarammdantas let me know what you're looking for feedback on.  I notice there's a number of TODOs still present here.

clarammdantas commented:
@timabbott for now, I was only in doubt if I really needed to add that `Web Public` options on stream creation. Because I saw a channel, `Rome`, in the dev environment that is already `Web Public`, so I was wondering if the flow for creating a `web public` stream is different and I didn't know about it. 

There are some TODOs that I put to remember places where I could possibly improve code, so I can go back and think about it, and the others you did and I haven't implemented yet.

I was also thinking why aren't the privacy modes an `Enum` instead of a `boolean`?

shubhamdhama commented:
> I was also thinking why aren't the privacy modes an Enum instead of a boolean?

+1 for this question, it should be `Model.PositiveSmallIntegerField` field IMO.

> @timabbott for now, I was only in doubt if I really needed to add that `Web Public` options on stream creation. Because I saw a channel, `Rome`, in the dev environment that is already `Web Public`, so I was wondering if the flow for creating a `web public` stream is different and I didn't know about it.

if I understand correctly you are asking for how to make a stream web-public? I think it's not implemented yet. For such cases, I follow unconventional ways, like writing a temporary script, management command, or simply using the IPython shell.

showell commented on static/js/stream_create.js:
```diff
@@ -150,18 +150,28 @@ function create_stream() {
     const privacy_setting = $('#stream_creation_form input[name=privacy]:checked').val();
 
     if (privacy_setting === 'invite-only') {
+        is_web_public = false;
```
This is probably fine for now, but we might want a follow up commit later that only sends fields that are really necessary.  I think we can say `invite_only == True` implies `is_web_public == False` on the server.

And then later:

~~~
if (is_web_public !== undefined) {
   data.is_web_public = JSON.stringify(is_web_public):
}
if (invite_only !== undefined) {
   // etc.

showell commented on static/js/stream_data.js:
```diff
@@ -872,7 +871,7 @@ exports.initialize = function (params) {
         data structures.
     */
-
+    // Aqui que eu encontro as web_public e todas as outras streams
```
Ingles, por favor. :)  Which bucket do the web-public subscriptions go into?

clarammdantas replied:
I'm sorry, @showell ! I pushed this yesterday just because it was easier to compare my changes after I rebased with master. I was trying to find out why I was getting that error I mentioned about the spinner on subscriptions, that turned out to be introduced in another commit.

I should have left a comment saying that and forgot, I will clean it up :)
Thanks for the review!

showell commented on static/js/subs.js:
```diff
@@ -781,16 +779,21 @@ function hide_subscribe_toggle_spinner(row_object) {
     let true_stream_name;
 
     if (row_object !== undefined) {
+        console.log(row_object);
```
Remove this eventually, please.

showell commented on static/js/subs.js:
```diff
@@ -781,16 +779,21 @@ function hide_subscribe_toggle_spinner(row_object) {
+
+    const data = {};
+    data.subscriptions = JSON.stringify([{name: stream, color: color}]);
+    data.is_web_public = JSON.stringify(is_web_public);
```
Should this flag be for each subscription?

clarammdantas replied:
I think yes because I removed the `@require_non_guest_user` decorator of the `add_subscriptions_backend` function making it possible for a guest user to subscribe to a stream. So I'm using this flag to check if the user has permission to subscribe, I decided to pass this flag in the request so I don't have to query the db for the stream to check only if it is `web public`.

aero31aero replied:
> I decided to pass this flag in the request so I don't have to query the db for the stream to check only if it is web public.

I haven't read the actual code below yet, but a malicious actor could easily send an API request with `data.is_web_public=true` for any stream and get subscribed. We should not trust data from the frontend or api for things we can check from the database.

aero31aero replied:
Overall, I'm against this particular change. We should do all the sanity checks on the backend itself.

clarammdantas replied:
Ok, so I've made a couple of manual tests for this. I tried to modify the HTTP request and pass the `is_web_pulbic = true` (as a normal user trying to subscribe to a private stream and as a guest user trying to subscribe to a public stream) and it didn't work. Then I hardcoded `is_web_public = true` on the `ajaxSubscribe` function and it didn't work as well. I also hardcoded `is_web_public = True` when it enters the `add_subscriptions_backend` function and it also didn't work. So, I guess that is no problem with this? @aero31aero @timabbott What do you think?

clarammdantas replied:
![Captura de tela de 2020-05-13 19-26-20](https://user-images.githubusercontent.com/15564640/81873022-42925f80-9551-11ea-9413-49a163011d88.png)

clarammdantas commented:
@shubhamdhama 

> if I understand correctly you are asking for how to make a stream web-public? I think it's not implemented yet. For such cases, I follow unconventional ways, like writing a temporary script, management command, or simply using the IPython shell.

I was just in doubt if that was implemented yet or not because I ended up implementing it, so I was afraid of being duplicating code. But it turns out it wasn't implemented yet, so everything is ok :)

timabbott commented:
> > I was also thinking why aren't the privacy modes an Enum instead of a boolean?
> +1 for this question, it should be Model.PositiveSmallIntegerField field IMO.

Agreed.  I think we should plan to change that in a follow-up PR; just to manage scope of this PR (especially since we already have the `is_web_public` field in the model).

timabbott commented:
This is looking good so far.  A comment on commit structure: I think this is best done as a relatively long series of commits that each do an individual, atomic, thing:
* A series of commits, each with tests, that change the various capabilities of guests for web_public streams on the backend.  E.g. one commit might be the `validate_user_access_to_subscribers_helper` code path.  
* A series of commits, with tests where the existing test suite supports it, that make the frontend work correctly for guest users across various components.
* Any pieces of my original commit that is just a reminder of things not yet done
* At the end, the commit to allow creating web_public streams in the API and then UI (We want to delay this until the feature works as described).  (A commit to support it in e.g. `make_stream` and `create_stream_if_needed` could come earlier as part of that first series).

And then we can incrementally review, finish, and merge the pieces of this that are finished, with tests, rather than accumulating a giant branch that has to be merged as a single chunk.

clarammdantas commented:
> This is looking good so far. A comment on commit structure: I think this is best done as a relatively long series of commits that each do an individual, atomic, thing:
> 
>     * A series of commits, each with tests, that change the various capabilities of guests for web_public streams on the backend.  E.g. one commit might be the `validate_user_access_to_subscribers_helper` code path.
> 
>     * A series of commits, with tests where the existing test suite supports it, that make the frontend work correctly for guest users across various components.
> 
>     * Any pieces of my original commit that is just a reminder of things not yet done
> 
>     * At the end, the commit to allow creating web_public streams in the API and then UI (We want to delay this until the feature works as described).  (A commit to support it in e.g. `make_stream` and `create_stream_if_needed` could come earlier as part of that first series).
> 
> 
> And then we can incrementally review, finish, and merge the pieces of this that are finished, with tests, rather than accumulating a giant branch that has to be merged as a single chunk.

Ok. I do prefer this way too, seeing things getting merged motivates me :)
Tomorrow I'm going to try to fix the commit structure. I guess I'm not very used to how to properly organize my commits. The way I was trying to do was make a separate commit for each flow like creating a web public stream, browse streams, and subscribe.

timabbott commented:
@clarammdantas just wanted to check in one whether you have any chunks ready to review/merge here rebased to the start of the branch :)

clarammdantas commented:
oh, I think most of the work is ready. I just would like to write a couple of tests I imagine would be good to have. But with these changes we already can:
1) Create a web public stream
2) Guest users can browse, subscribe and unsubscribe to web public streams
3) See in "All Streams" all web public streams, subscribed and unsubscribed streams

@timabbott 
I expect it to be fully ready for review on Thursday :)

aero31aero commented on static/js/stream_data.js:
```diff
@@ -448,17 +450,18 @@ exports.update_calculated_fields = function (sub) {
-    sub.can_access_subscribers = page_params.is_admin || sub.subscribed || !page_params.is_guest &&
-                                 !sub.invite_only;
+    // Guest users can only access subscribers of web_public streams they can join.
+    sub.can_access_subscribers = page_params.is_admin || sub.should_display_preview_button &&
```
This boolean is quickly getting unwieldy, we should break it down.

clarammdantas replied:
Yes, it was a pain dealing with all those booleans. There were some changes I did that I was afraid of not be doing what I expected, so I ended up doing the truth table just in case.

But after this get done, I was actually going to ask about the possibility of refactoring this and use Enums instead of booleans so it's easier to deal with all those conditions, and you can do changes that are correct.

aero31aero commented on static/js/stream_data.js:
```diff
@@ -892,7 +895,6 @@ exports.initialize = function (params) {
         We only use it in this function to populate other
         data structures.
     */
-
```
Stray extra line or a deliberate change for readability?

aero31aero commented on static/js/stream_create.js:
```diff
@@ -150,18 +150,28 @@ function create_stream() {
+    let is_web_public;
     const privacy_setting = $('#stream_creation_form input[name=privacy]:checked').val();
 
+    // TODO: Infeer the variables so there is no need to define all three.
```
Typo in the TODO. Infeer=>Infer

aero31aero commented on static/js/stream_create.js:
```diff
@@ -150,18 +150,28 @@ function create_stream() {
+        is_web_public = false;
         invite_only = true;
         history_public_to_subscribers = true;
+    } else if (privacy_setting === 'is-web-public') {
```
This whole bit of code is ripe for some refactor. We shouldn't have to specify `is_web_public=false` 3 times.

```js
data.is_web_public = ['is-web-public'].includes(prvacy_setting);
data.history_public_to_subscribers = ['is-web-public', 'public', 'invite-only-public-history'].includes(privacy_setting);
data.invite_only = ['invite-only-public-history', 'invite-only'].includes(privacy_setting);
```

aero31aero replied:
This can go into a commit of its own, if you want to make this change now.

clarammdantas replied:
Yes, I leave that TODO to remind me of working on this, so if a stream is `web_public` we just need to pass the `web_public` arguments because of we now for this case `invite_only = False`, `history_public_to_subscribers = True`. There is no much need to define those on the frontend I guess, just pass the relevant values to the back and let it deal with it.

aero31aero replied:
There is certain advantage in computing this here and telling the API what to do than having the API try to figure out these parameters.

For example, there isn't a direct field in the database for `invite-only-public-history`; its a combination of two flags.

clarammdantas replied:
Yeah, but for that case, I would only set the two values instead of three. But is easier to fix it how you showed :)

aero31aero commented on zerver/lib/actions.py:
```diff
@@ -1834,7 +1834,6 @@ def check_send_typing_notification(sender: UserProfile, notification_to: Union[S
         operator=operator,
     )
 
-
```
Stray newline. :)

clarammdantas replied:
Oh, I deleted this because there was 2 blank lines between this method and the one below, and isn't Zulip convention to separate methods with one blank line?

aero31aero commented on zerver/lib/streams.py:
```diff
@@ -47,10 +47,12 @@ def send_stream_creation_event(stream: Stream, user_ids: List[int]) -> None:
                             stream_post_policy: int=Stream.STREAM_POST_POLICY_EVERYONE,
                             history_public_to_subscribers: Optional[bool]=None,
                             stream_description: str="") -> Tuple[Stream, bool]:
+
```
Same issue with a stray newline here.

aero31aero commented on zerver/tests/test_subs.py:
```diff
@@ -359,6 +366,30 @@ def test_make_stream_public(self) -> None:
+            'history_public_to_subscribers': ujson.dumps(True)
+        }
+        stream = get_stream('web_public_stream', realm)
+        result = self.client_patch("/json/streams/%d" % (stream.id,), params)
```
We should try to use f-strings in the newer code now that we support Python 3.6.

aero31aero commented on frontend_tests/node_tests/stream_data.js:
```diff
@@ -475,27 +475,42 @@ run_test('stream_settings', () => {
     assert.equal(sub_rows[0].color, 'blue');
-    assert.equal(sub_rows[1].color, 'amber');
-    assert.equal(sub_rows[2].color, 'cinnamon');
+    assert.equal(sub_rows[1].color, 'red');
```
We could potentially make this code more readable by using `assert.equal(sub_rows[1].color, web_public_stream.color);` and similar.

clarammdantas replied:
Ok. I did a preliminary commit changing the others.

aero31aero commented on frontend_tests/node_tests/stream_data.js:
```diff
@@ -509,14 +524,138 @@ run_test('stream_settings', () => {
 
-    // For guest user only retrieve subscribed streams
+    // For guest user retrieve subscribed, unsubscribed and never subscribed
+    // streams. The unsubscribed streams are shown but unless they are web
```
We could reword this to be:

"For guest users, retrieve:
- subscribed streams
- previously subscribed streams
- all web_public streams"

YashRE42 commented on static/js/stream_create.js:
```diff
@@ -148,29 +148,11 @@ function create_stream() {
-        history_public_to_subscribers = true;
-    }
+    const is_web_public = ['is-web-public'].includes(privacy_setting);
+    const invite_only = ['is-web-public', 'public', 'invite-only-public-history'].includes(privacy_setting);
```
This case:
```
if (privacy_setting === 'is-web-public') {	
        is_web_public = true;	
        invite_only = false;	
        history_public_to_subscribers = true;	
}
```
and this case:
```
invite_only = ['is-web-public', 'public', 'invite-only-public-history'].includes(privacy_setting)
```

Are inconsistent.... That looks like a mistake to me

YashRE42 commented on static/js/stream_create.js:
```diff
@@ -148,29 +148,11 @@ function create_stream() {
-    }
+    const is_web_public = ['is-web-public'].includes(privacy_setting);
+    const invite_only = ['is-web-public', 'public', 'invite-only-public-history'].includes(privacy_setting);
+    const history_public_to_subscribers = ['invite-only-public-history', 'invite-only'].includes(privacy_setting);
```
Aah. I think you just got `history_public_to_subscribers` and `invite_only` mixed up in this commit.

clarammdantas commented:
@timabbott this is ready for review

timabbott commented on zerver/lib/actions.py:
```diff
@@ -4979,7 +4987,7 @@ def gather_subscriptions_helper(user_profile: UserProfile,
         never_subscribed_stream_ids = all_streams_id_set - sub_unsub_stream_ids
     else:
-        never_subscribed_stream_ids = set()
+        never_subscribed_stream_ids = set(web_public_stream_ids)
```
Won't this potentially result in duplicates?  I think we probably want to change this to be `set(web_public_stream_ids - sub_unsubs_stream_ids`, right?

timabbott commented on zerver/tests/test_subs.py:
```diff
@@ -3959,6 +3959,10 @@ def test_guest_user_case() -> None:
 
+            for stream_dict in never_subscribed:
+                self.assertTrue(stream_dict["is_web_public"])
+                self.assertEqual(len(stream_dict["subscribers"]), len(users_to_subscribe))
```
The first assertion here belongs in the previous commit, right?  

And an assertion that everything in `never_subscribed` is not in the other lists might also be valuable, since that's the bug the first commit has.

timabbott commented on zerver/tests/test_subs.py:
```diff
@@ -4127,6 +4131,33 @@ def test_json_get_subscribers(self) -> None:
+            subscriber = UserProfile.objects.get(email=subscriber)
+            subscribers.append(subscriber.id)
+
+        self.assertEqual(set(subscribers), set(expected_subscribers))
```
We should make sure we assert the list isn't empty somewhere in this test; I might be happy to just do that (replacing the whole expected list generation dance).

timabbott commented on zerver/lib/actions.py:
```diff
@@ -2584,6 +2585,10 @@ def validate_user_access_to_subscribers_helper(
+    # Even guest users can join web-public streams
+    if stream_dict["is_web_public"]:
+        return
+
```
This change lgtm.

timabbott commented on zerver/tests/test_message_send.py:
```diff
@@ -1885,6 +1885,18 @@ def test_basic_check_message_call(self) -> None:
+        # Guest users can write to web_public streams.
+        sender = self.example_user("polonius")
+        client = make_client(name="test suite")
+        stream_name = "Web Public Stream"
```
Can we just use Rome here, with an assertion that Polonium isn't subscribed and that it's web_public, rather than making a new stream?

timabbott commented on zerver/lib/streams.py:
```diff
@@ -163,6 +163,10 @@ def access_stream_for_send_message(sender: UserProfile,
             raise JsonableError(_("New members cannot send to this stream."))
 
+    if stream.is_web_public:
+        # Guest users can write to web_public streams
```
s/Guest users/Even guest users/ here.

timabbott commented on zerver/lib/streams.py:
```diff
@@ -408,6 +408,10 @@ def filter_stream_authorization(user_profile: UserProfile,
+            continue
+
         # Users are not authorized for invite_only streams, and guest
         # users are not authorized for any streams
```
Let's update this comment and maybe change this next block in a preparatory refactoring to be:
```
# Members and administrators are authorized for public streams
if not stream.invite_only and not user_profile.is_guest:
    continue

unauthorized_streams.append(stream)
```

(I just think the flow will be better with all the cases being `continue`).

timabbott commented on zerver/views/streams.py:
```diff
@@ -597,13 +597,15 @@ def get_subscribers_backend(request: HttpRequest, user_profile: UserProfile,
 def get_streams_backend(
         request: HttpRequest, user_profile: UserProfile,
         include_public: bool=REQ(validator=check_bool, default=True),
+        include_web_public: bool=REQ(validator=check_bool, default=True),
```
I think this should perhaps default to `False`; they should already be included through `include_public`, and I don't want to make it more complicated to ask for "just my subscriptions" (as this change visibly does in the API documentation above).

timabbott commented:
@clarammdantas this is really great progress!  The rebased version, at least with its first ~5 commits, is exactly what I was looking for in order to review this efficiently.  I merged the first commit as 831298d387ba09856c50139617644a5f436d464f and posted a batch of comments on the next ~5.  Let me know when those are resolved and I'll continue incrementally merging a prefix of this PR.

clarammdantas commented:
@timabbott thanks!! This makes me really happy! :)

I've fixed all the things suggested and added an extra commit with changes that I think would made `test_subs` more readable. It's ready for a new review.

timabbott commented:
For your first commit, can you use the `urllib` library to turn a dictionary of parameters into the URL without all the messy line-breaking logic.  It's be more idiomatically correct as well as cleaner.

timabbott commented:
I applied the following change to `validate_user_access_to_subscribers_helper`.  The first hunk, beyond being a comment fixup, improves performance slightly by avoiding checking whether the user is a subscriber when it's a web-public stream and that detail doesn't matter.

```
diff --git a/zerver/lib/actions.py b/zerver/lib/actions.py
index 458b53b576..4b4bcd744b 100644
--- a/zerver/lib/actions.py
+++ b/zerver/lib/actions.py
@@ -2577,6 +2577,11 @@ def validate_user_access_to_subscribers_helper(
     if user_profile.realm_id != stream_dict["realm_id"]:
         raise ValidationError("Requesting user not in given realm")
 
+    # Even guest users can access subscribers to web-public streams,
+    # since they can freely become subscribers to these streams.
+    if stream_dict["is_web_public"]:
+        return
+
     # Guest users can access subscribed public stream's subscribers
     if user_profile.is_guest:
         if check_user_subscribed(user_profile):
@@ -2585,10 +2590,6 @@ def validate_user_access_to_subscribers_helper(
         # any code paths that would allow a guest user to access other
         # streams in the first place.
 
-    # Even guest users can join web-public streams
-    if stream_dict["is_web_public"]:
-        return
-
     if not user_profile.can_access_public_streams() and not stream_dict["invite_only"]:
         raise JsonableError(_("Subscriber data is not available for this stream"))
 
diff --git a/zerver/tests/test_subs.py b/zerver/tests/test_subs.py
index 0398506341..55561c87fc 100644
--- a/zerver/tests/test_subs.py
+++ b/zerver/tests/test_subs.py
@@ -3967,9 +3967,11 @@ class GetSubscribersTest(ZulipTestCase):
                 self.assertTrue(stream_dict["stream_id"] not in sub_ids)
                 self.assertTrue(stream_dict["stream_id"] not in unsub_ids)
 
-                if stream_dict["name"] != "Rome":
-                    # Rome is a already created and have other subscribers than
-                    # the ones created in this test.
+                # The Rome stream has is_web_public=True, and already
+                # created and has default subscribers not included in
+                # this test, so we do the following check only for the
+                # streams we created.
+                if stream_dict["name"] in web_public_streams:
                     self.assertEqual(
                         len(stream_dict["subscribers"]),
                         len(users_to_subscribe))
```

timabbott commented on zerver/models.py:
```diff
@@ -2213,6 +2213,18 @@ def validate_attachment_request(user_profile: UserProfile, path_id: str) -> Opti
+            was_sent_to_web_public_stream):
+        # Any user in the realm, including guest users, can access files sent
+        # to web-public streams.
+        return True
```
This is a somewhat got code path, so we should be careful about adding more complexity to it.

I think we might be able to do this with fewer database queries by combining the logic with the next check.

It's also worth thinking about whether we want to add an `attachment.is_web_public` to sit alongside `is_realm_public`; that might be a better way to hook things up long-term, since when we add the main "is_web_public" feature of having content be public to the web, we might want to make it very cheap to do that check.

clarammdantas replied:
I think adding `is_web_public` to the `attachment` is a better way to do it too. Could I go ahead and implement it? Or this would need more discussion?

timabbott replied:
I think you should go for it.  Might be reasonable to just do that as an independent change on a new PR to make it easy to have CI run on it (I expect it'll involve some back-and-forth discussion).

timabbott commented:
I merged about half of this as the series ending with c3779338c6fde0c6f3c3307a21d9b6eef23ffd9d, see the notes above.  @amanagr @mateuszmandera FYI.

timabbott commented:
There should probably be a commit somewhat earlier in this PR that adds `is_web_public` to Stream.API_FIELDS?  I'm not sure precisely where.

timabbott commented on static/js/stream_create.js:
```diff
@@ -150,20 +150,17 @@ function create_stream() {
+        "invite-only-public-history",
+        "is-web-public",
+        "public",
+    ].includes(privacy_setting);
```
I think the rewrite to use `includes` is probably less readable than the old logic; I'd change this to just do it the verbose way.

timabbott commented on static/templates/stream_types.hbs:
```diff
@@ -8,6 +8,12 @@
+        <label class="radio">
+            <input type="radio" name="privacy" value="is-web-public" {{#if is_web_public}}checked{{/if}}/>
+            {{t '<b>Web public:</b> anyone (including guest users) can join; can view complete message history without joining' }}
+        </label>
```
This should say clearly that anyone on the Internet can access complete message history without creating an account.  We may defer merging this commit until we've made that the case.

andersk commented:
I reverted “digest: Support digest of web public streams for guest users” because it was failing CI without the future commit “streams: Create web public streams”, which adds the `is_web_public` parameter to `create_stream_if_needed`. This will need a bit of rearranging.

clarammdantas commented:
> I reverted “digest: Support digest of web public streams for guest users” because it was failing CI without the future commit “streams: Create web public streams”, which adds the `is_web_public` parameter to `create_stream_if_needed`. This will need a bit of rearranging.

Sorry!! :x
I rearraged the commits. Nice commit message BTW hahaha

clarammdantas commented on zerver/lib/actions.py:
```diff
@@ -5017,11 +5017,11 @@ def gather_subscriptions_helper(user_profile: UserProfile,
-                subscribers = subscriber_map[stream["id"]]
-                if subscribers is not None:
-                    stream_dict['subscribers'] = subscribers
+            subscribers = subscriber_map[stream["id"]]
```
This is already inside of a `if is_public or user_profile.is_realm_admin:` condition, line 5003, so I think it can be removed. I ran the `test_subs` test after this removal and it was fine.

timabbott commented:
I merged the first 2 commits; I think the next thing to do is finish the attachments field work and then we can come back to this.

timabbott commented:
@clarammdantas can you rebase this?  Now that the attachments commit is done, we should try to focus on this PR.

clarammdantas commented:
@timabbott This is ready for a new review.

timabbott commented on zerver/tests/test_subs.py:
```diff
@@ -397,6 +397,30 @@ def test_make_stream_public(self) -> None:
+        self.make_stream('web_public_stream',
+                         realm=realm,
+                         is_web_public=True,
+                         history_public_to_subscribers=True)
```
I don't understand this test; it seems like the parameters you set via the API (below) match how you created it?

clarammdantas replied:
I wanted to check if a user can create a web-public stream and if it was accessible/editable to other users as well.

hmm, I guess just checking if it was created right is better?

timabbott replied:
Yeah, exactly, I'd just do `Stream.objects.get()` if the goal is to verify creation happened successfully.

timabbott replied:
(Is the intent to use `create_stream_if_needed`?)

timabbott commented on zerver/openapi/zulip.yaml:
```diff
@@ -4600,6 +4600,14 @@ paths:
+        - name: is_web_public
+          in: query
+          description: A boolean specifying whether the streams specified in
+            `subscriptions` are web-public or not.
```
Can you follow the pattern for `invite_only` is described (above)?

timabbott commented on zerver/lib/digest.py:
```diff
@@ -148,9 +148,14 @@ def gather_hot_conversations(user_profile: UserProfile, messages: List[Message])
             invite_only=False, date_created__gt=threshold))
+
     else:
         new_streams = []
```
I think this new else clause is impossible; we should just be replacing this line with the `is_guest` case.

is_guest is equivalent to `!can_access_public_streams()`.

clarammdantas replied:
Indeed.

timabbott commented:
@clarammdantas I posted a batch of comments that once resolved should let us merge the next ~3 commits.

timabbott commented:
@clarammdantas let me know when you have this ready again!  I'm hoping we can integrate this feature for the 4.0 release.

clarammdantas commented:
@timabbott  sure! I'll try to get it done by tomorrow morning.

timabbott commented on zerver/lib/streams.py:
```diff
@@ -113,6 +115,7 @@ def create_streams_if_needed(realm: Realm,
             realm,
             stream_dict["name"],
             invite_only=stream_dict.get("invite_only", False),
+            is_web_public=stream_dict.get("is_web_public", False),
```
So the test included in this commit doesn't actually test this code path; is that intentional?

timabbott commented:
I merged 867428719245e35bb4493295bed5282d50a9e1ab after looking at this for a bit; I'd encourage you to reread your commit messages and changes after rebasing this next.  I think the next commit we should be able to merge is "streams: Allow guest users to subscribe themselves." so please polish that one -- the commit message is a bit misformatted.

Thanks @clarammdantas!

timabbott commented:
@clarammdantas ping on rebasing this!

timabbott commented:
@amanagr maybe you should pick up doing a quick rebase on this since it seems @clarammdantas is busy?  I think it only has a bit more work and connects closely to the `is_web_public` PR that you just integrated.

As a summary of the strategy here -- I'd like to integrate all the changes for what an `is_web_public` stream does, but not the final commit (currently first in the series) that makes it possible to actually create `is_web_public` streams in the UI (basically because we want to explicitly review the security story for this previously impossible case before doing that).

@sahil839 FYI as well in case you want to coordinate with Aman in trying to get this in.

amanagr commented:
Okay, Happy to finish this work.

amanagr commented:
Closing in favour of #16728
